src/Castle.Core.Tests/BugsReported/Core293.cs

```diff
@@ -0,0 +1,112 @@
+﻿// Copyright 2004-2017 Castle Project - http://www.castleproject.org/
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+namespace Castle.DynamicProxy.Tests.BugsReported
+{
+	using System;
+
+	using NUnit.Framework;
+
+	public class LazyInterceptor<T> : IInterceptor
+		where T : class
+	{
+		public LazyInterceptor(Lazy<T> lazyTarget)
+		{
+			LazyTarget = lazyTarget;
+		}
+
+		private Lazy<T> LazyTarget { get; }
+
+		public void Intercept(IInvocation invocation)
+		{
+			var target = invocation.InvocationTarget as T;
+			if (target == null)
+			{
+				(invocation as IChangeProxyTarget).ChangeInvocationTarget(LazyTarget.Value);
+				(invocation as IChangeProxyTarget).ChangeProxyTarget(LazyTarget.Value);
+			}
+
+			invocation.Proceed();
+		}
+	}
+
+	public interface IEventHandler
+	{
+		string Handle(EventArgs ea);
+	}
+
+	public interface IEventHandler<T> : IEventHandler
+		where T : EventArgs
+	{
+	}
+
+	public class EventArgs1 : EventArgs { }
+
+	public class Handler1 : IEventHandler<EventArgs1>
+	{
+		public string Handle(EventArgs ea)
+		{
+			return "Handler1";
+		}
+	}
+
+	public class EventArgs2 : EventArgs { }
+
+	public class Handler2 : IEventHandler<EventArgs2>
+	{
+		public string Handle(EventArgs ea)
+		{
+			return "Handler2";
+		}
+	}
+
+	public interface IEventHandler3 : IEventHandler { }
+
+	public class Handler3 : IEventHandler3
+	{
+		public string Handle(EventArgs ea)
+		{
+			return "Handler3";
+		}
+	}
+
+	[TestFixture]
+	public class Core293 : BasePEVerifyTestCase
```

Comment:
That BugsReported directory contains tests relating to JIRA issues and is nice and old. It would be great if you could move it into `src/Castle.Core.Tests/DynamicProxy.Tests` with a descriptive name, `ChangeProxyTargetTestCase` maybe.